clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -0,0 +1,491 @@
+    ComputeOutSets(EB, BlockMap, WorkList);
+  }
+
+  CollectWidenedBounds(BlockMap);
+}
+
+void BoundsAnalysis::ComputeGenSets(BlockMapTy BlockMap) {
+  // If there is an edge B1->B2 and the edge condition is of the form
+  // "if (*(p + i))" then Gen[B1] = {B2, p:i} .
+
+  for (const auto B : BlockMap) {
+    auto EB = B.second;
+
+    for (const CFGBlock *pred : EB->Block->preds()) {
+      if (SkipBlock(pred))
+        continue;
+
+      // We can add "p:i" only on the true edge.
+      // For example,
+      // B1: if (*(p + i))
+      // B2:   foo();
+      // B3: else bar();
+
+      // Here we have the edges (B1->B2) and (B1->B3). We can add "p:i" only
+      // on the true edge. Which means we will add the following entry to
+      // Gen[B1]: {B2, p:i}
+      if (const auto *I = pred->succs().begin())
+        if (*I != EB->Block)
+          continue;
+
+      // Get the edge condition and fill the Gen set.
+      if (Expr *E = GetTerminatorCondition(pred))
+        FillGenSet(E, BlockMap[pred], EB);
+    }
+  }
+}
+
+void BoundsAnalysis::CollectBoundsVars(const Expr *E, DeclSetTy &BoundsVars) {
+  if (!E)
+    return;
+
+  E = IgnoreCasts(const_cast<Expr *>(E));
+
+  // Collect bounds vars for the lower and upper bounds exprs.
+  // Example:
+  // _Nt_array_ptr<char> p : bounds(p + i, p + j);
+  // LowerExpr: p + i.
+  // UpperExpr: p + j.
+  if (const auto *RBE = dyn_cast<RangeBoundsExpr>(E)) {
+    CollectBoundsVars(RBE->getLowerExpr(), BoundsVars);
+    CollectBoundsVars(RBE->getUpperExpr(), BoundsVars);
+  }
+
+  // Collect bounds vars for the LHS and RHS of bounds exprs.
+  // Example: In p + i, LHS = p and RHS = i
+  if (const auto *BO = dyn_cast<BinaryOperator>(E)) {
+    CollectBoundsVars(BO->getLHS(), BoundsVars);
+    CollectBoundsVars(BO->getRHS(), BoundsVars);
+  }
+
+  if (const auto *D = dyn_cast<DeclRefExpr>(E)) {
+    if (const auto *V = dyn_cast<VarDecl>(D->getDecl()))
+      BoundsVars.insert(V);
+  }
+}
+
+void BoundsAnalysis::FillGenSet(Expr *E,
+                                ElevatedCFGBlock *EB,
+                                ElevatedCFGBlock *SuccEB) {
+
+  // Handle if conditions of the form "if (*e1 && *e2)".
+  if (const auto *BO = dyn_cast<const BinaryOperator>(E)) {
+    if (BO->getOpcode() == BO_LAnd) {
+      FillGenSet(BO->getLHS(), EB, SuccEB);
+      FillGenSet(BO->getRHS(), EB, SuccEB);
+    }
+  }
+
+  // Check if the edge condition contains a pointer deref.
+  if (!ContainsPointerDeref(E))
+    return;
+
+  E = IgnoreCasts(E);
+
+  if (const auto *UO = dyn_cast<UnaryOperator>(E)) {
+    const auto *Exp = IgnoreCasts(UO->getSubExpr());
+    if (!Exp)
+      return;
+
+    // TODO: Handle accesses of the form:
+    // "if (*(p + i) && *(p + j) && *(p + k))"
+
+    // For conditions of the form "if (*p)".
+    if (const auto *D = dyn_cast<DeclRefExpr>(Exp)) {
+      if (const auto *V = dyn_cast<VarDecl>(D->getDecl())) {
+        if (V->getType()->isCheckedPointerNtArrayType()) {
```

Comment:
There's some confusion about the meaning of the gen entries.  I think you want gen to be a dereference expression, not a constant offset.  When you widen bounds, you need to check that the dereference is at exactly the upper bound of the variable.

Currently gen entries are  constant offsets.  The question is what is an entry a  constant offset from.   If it is from a variable, then we can't handle cases where the upper bound constant has the from v + i, where v is a variable and i is a variable too.    If it is from the upper bound of v, that creates a problem, because the upper bound can change if the bounds of v are widened.

Reply:
I agree that gen can remain a constant offset.  However, it should be a constant offset relative to the declared upper bound of a pointer variable.